eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxos.hs

```diff
@@ -1,59 +1,143 @@
 
+import Cardano.Ledger.Alonzo.Language (Language (..))
 import Cardano.Ledger.Alonzo.Rules (
   AlonzoUtxoEvent (..),
   AlonzoUtxoPredFailure (..),
-  AlonzoUtxosEvent,
-  AlonzoUtxosPredFailure,
+  AlonzoUtxosEvent (..),
+  AlonzoUtxosPredFailure (..),
+  validBegin,
+  validEnd,
  )
 import Cardano.Ledger.Alonzo.Scripts (AlonzoScript)
-import Cardano.Ledger.Alonzo.Tx (AlonzoTx (..))
-import Cardano.Ledger.Babbage.Rules (BabbageUTXO, BabbageUtxoPredFailure (..))
+import Cardano.Ledger.Alonzo.TxInfo (EraPlutusContext, ExtendedUTxO)
+import Cardano.Ledger.Alonzo.UTxO (AlonzoScriptsNeeded)
+import Cardano.Ledger.Babbage.Rules (BabbageUTXO, BabbageUtxoPredFailure (..), expectScriptsToPass, scriptsNo, tellDepositChangeEvent)
+import Cardano.Ledger.Babbage.Tx
 import Cardano.Ledger.BaseTypes (ShelleyBase)
 import Cardano.Ledger.Conway.Core
 import Cardano.Ledger.Conway.Era (ConwayUTXOS)
-import Cardano.Ledger.Shelley.LedgerState (PPUPPredFailure, UTxOState (..))
-import Cardano.Ledger.Shelley.Rules (UtxoEnv (..))
-import Control.State.Transition.Extended (Embed (..), STS (..))
+import Cardano.Ledger.Conway.Governance (ConwayGovernance (..))
+import Cardano.Ledger.Shelley.LedgerState (
+  PPUPPredFailure,
+  UTxOState (..),
+ )
+import Cardano.Ledger.Shelley.Rules (
+  UtxoEnv (..),
+  updateUTxOState,
+ )
+import Cardano.Ledger.UTxO (EraUTxO (..))
+import Control.State.Transition.Extended
+import Debug.Trace (traceEvent)
+import Lens.Micro
 
 instance
-  ( EraTxOut era
+  ( AlonzoEraTx era
+  , AlonzoEraPParams era
+  , BabbageEraTxBody era
+  , EraTxOut era
   , EraGovernance era
   , EraTxCert era
+  , EraUTxO era
+  , ExtendedUTxO era
+  , EraPlutusContext 'PlutusV1 era
+  , GovernanceState era ~ ConwayGovernance era
   , Script era ~ AlonzoScript era
+  , ScriptsNeeded era ~ AlonzoScriptsNeeded era
+  , Tx era ~ AlonzoTx era
   , Eq (PPUPPredFailure era)
   , Show (PPUPPredFailure era)
   ) =>
   STS (ConwayUTXOS era)
   where
-  type BaseM (ConwayUTXOS era) = ShelleyBase
+  type BaseM (ConwayUTXOS era) = Cardano.Ledger.BaseTypes.ShelleyBase
   type Environment (ConwayUTXOS era) = UtxoEnv era
   type State (ConwayUTXOS era) = UTxOState era
   type Signal (ConwayUTXOS era) = AlonzoTx era
   type PredicateFailure (ConwayUTXOS era) = AlonzoUtxosPredFailure era
   type Event (ConwayUTXOS era) = AlonzoUtxosEvent era
 
-  transitionRules = []
+  transitionRules = [utxosTransition]
 
 instance
-  ( EraTxOut era
+  ( AlonzoEraTx era
+  , BabbageEraTxBody era
   , EraGovernance era
+  , EraPlutusContext 'PlutusV1 era
+  , EraTxOut era
   , EraTxCert era
-  , PredicateFailure (EraRule "UTXOS" era) ~ AlonzoUtxosPredFailure era
+  , EraUTxO era
+  , ExtendedUTxO era
   , Event (EraRule "UTXOS" era) ~ AlonzoUtxosEvent era
+  , GovernanceState era ~ ConwayGovernance era
+  , PredicateFailure (EraRule "UTXOS" era) ~ AlonzoUtxosPredFailure era
+  , Script era ~ AlonzoScript era
+  , ScriptsNeeded era ~ AlonzoScriptsNeeded era
+  , Tx era ~ AlonzoTx era
```

Comment:
```suggestion
```

Reply:
Same as above, I could only remove it if adding  `Signal (ConwayUTXOS era) ~ Tx era `